internal/integration/integration_test.go

```diff
@@ -30,76 +29,106 @@ func TestIntegration(t *testing.T) {
+		}
+
+		// get <cluster>-talos secret
 		var talosSecret corev1.Secret
 		key = types.NamespacedName{
 			Namespace: namespaceName,
 			Name:      cluster.Name + "-talos",
 		}
 		require.NoError(t, c.Get(ctx, key, &talosSecret))
-		assert.Len(t, talosSecret.Data, 3)
-		assert.NotEmpty(t, talosSecret.Data["certs"]) // TODO more tests
-		assert.NotEmpty(t, talosSecret.Data["kubeSecrets"])
-		assert.NotEmpty(t, talosSecret.Data["trustdInfo"])
 
+		// check <cluster>-talos secret
+		{
+			assert.Len(t, talosSecret.Data, 3)
+			assert.NotEmpty(t, talosSecret.Data["certs"])
+			assert.NotEmpty(t, talosSecret.Data["kubeSecrets"])
+			assert.NotEmpty(t, talosSecret.Data["trustdInfo"])
+		}
+
+		// get <machine>-bootstrap-data secret
 		var bootstrapDataSecret corev1.Secret
 		key = types.NamespacedName{
 			Namespace: namespaceName,
 			Name:      machine.Name + "-bootstrap-data",
 		}
 		require.NoError(t, c.Get(ctx, key, &bootstrapDataSecret))
-		assert.Len(t, bootstrapDataSecret.Data, 1)
-		provider, err := configloader.NewFromBytes(bootstrapDataSecret.Data["value"])
-		require.NoError(t, err)
 
-		provider.(*v1alpha1.Config).ClusterConfig.ControlPlane.Endpoint.Host = "FIXME"
+		// check <machine>-bootstrap-data secret
+		var provider machineconfig.Provider
+		{
+			assert.Len(t, bootstrapDataSecret.Data, 1)
+			var err error
+			provider, err = configloader.NewFromBytes(bootstrapDataSecret.Data["value"])
+			require.NoError(t, err)
+			_, err = provider.Validate(runtimeMode{false}, machineconfig.WithStrict())
+			require.NoError(t, err)
+		}
+
+		// cross-checks
+		{
+			secretsBundle := generate.NewSecretsBundleFromConfig(generate.NewClock(), provider)
+
+			var certs generate.Certs
+			require.NoError(t, yaml.Unmarshal(talosSecret.Data["certs"], &certs))
+			assert.NotEmpty(t, certs.Admin)
+			certs.Admin = nil
+			assert.Equal(t, secretsBundle.Certs, &certs)
+			assert.Equal(t, caSecret.Data[corev1.TLSCertKey], certs.K8s.Crt)
+
+			var kubeSecrets generate.Secrets
+			require.NoError(t, yaml.Unmarshal(talosSecret.Data["kubeSecrets"], &kubeSecrets))
+			assert.Equal(t, secretsBundle.Secrets, &kubeSecrets)
+
+			var trustdInfo generate.TrustdInfo
+			require.NoError(t, yaml.Unmarshal(talosSecret.Data["trustdInfo"], &trustdInfo))
+			assert.Equal(t, secretsBundle.TrustdInfo, &trustdInfo)
+		}
+
+		// create the second machine
+		machine2 := createMachine(ctx, t, c, cluster)
+		talosConfig2 := createTalosConfig(ctx, t, c, machine2)
+		waitForReady(ctx, t, c, talosConfig2)
+
+		// get <machine>-bootstrap-data secret
+		var bootstrapDataSecret2 corev1.Secret
+		key = types.NamespacedName{
+			Namespace: namespaceName,
+			Name:      machine2.Name + "-bootstrap-data",
+		}
+		require.NoError(t, c.Get(ctx, key, &bootstrapDataSecret2))
 
-		// TODO more tests
-		_, err = provider.Validate(runtimeMode{false}, config.WithStrict())
-		require.NoError(t, err)
+		assert.Equal(t, bootstrapDataSecret.Data, bootstrapDataSecret2.Data) // ?!
```

Comment:
interesting approach, but looks good to me. I bet with Talos 0.12 machinery this _should_ probably break